### PR TITLE
Edit user addresses

### PR DESCRIPTION
**Pull request by bingxie**
User can edit addresses on user account page. #1131

### CODE REVIEW

File: spec/features/consumer/account_spec.rb [hunk 1]
```diff
@@ -60,4 +60,33 @@
...
+      fill_in('user[ship_address_attributes][firstname]', with: 'John')
+      fill_in('user[ship_address_attributes][lastname]', with: 'Doe')
+      fill_in('user[ship_address_attributes][address1]', with: '11 Lovely Street')
+      fill_in('user[ship_address_attributes][phone]', with: '123-456-7890')
+      fill_in('user[ship_address_attributes][city]', with: 'Merbourne')
+      fill_in('user[ship_address_attributes][zipcode]', with: '3127')
+
+      click_button('Update')
+
+      expect(page).to have_content '11 Lovely Street'
```
Comment (mkllnk) [inline spec/features/consumer/account_spec.rb]: Do you need to make sure that the data actually saved?

I'm just imagining an "Update" button that doesn't do anything or works very slowly. The next three lines would just test the currently filled out form that is not submitted yet.

File: app/views/spree/users/show.html.haml [hunk 2]
```diff
@@ -7,6 +7,34 @@
       .account-summary{"data-hook" => "account_summary"}
         = @user.email
         (#{link_to t(:edit), spree.edit_account_path})
+      %hr
+      %h3= t(:my_addresses)
```
Comment (mkllnk) [inline app/views/spree/users/show.html.haml]: When I did the internationalisation first, it was easy to write `t(:my_addresses)`. But in most views it's better to use translations in a local scope: `t('.my_addresses')`. The locale then needs an entry like:
```yml
en:
  spree:
    users:
      show:
        my_address: 'My Addresses'
```

This way you avoid any conflicts between views and the locale file is more structured. What do you think?

File: app/views/spree/users/edit_address.html.haml [hunk 3]
```diff
@@ -0,0 +1,15 @@
+.darkswarm
+  .row
+    = render :partial => 'spree/shared/error_messages', :locals => { :target => @user }
```
Comment (mkllnk) [inline app/views/spree/users/edit_address.html.haml]: Can't you use the new ruby syntax here? `render partial: '...', locals: {...}`

Comment (bingxie) [inline app/views/spree/users/edit_address.html.haml]: yes, sure!

File: app/views/spree/users/show.html.haml (hunk 2 above)
Comment (bingxie) [inline app/views/spree/users/show.html.haml]: very good suggestion!

File: spec/features/consumer/account_spec.rb (hunk 1 above)
Comment (bingxie) [inline spec/features/consumer/account_spec.rb]: good spot, will improve it.

**Comment by bingxie** [comment]
Just updated, could you have a look again @mkllnk ?

File: app/controllers/spree/users_controller_decorator.rb [hunk 4]
```diff
@@ -1,3 +1,21 @@
 Spree::UsersController.class_eval do
   layout 'darkswarm'
+
+  prepend_before_filter :load_object, :only => [:edit_address, :update_address, :show, :edit, :update]
```
Comment (mkllnk) [inline app/controllers/spree/users_controller_decorator.rb]: Still old syntax here. I use `rubocop` to check my code before committing.

File: app/controllers/spree/users_controller_decorator.rb [hunk 5]
```diff
@@ -1,3 +1,21 @@
...
+    country_id = Spree::Config[:default_country_id]
+    @user.build_ship_address(country_id: country_id) unless @user.ship_address
+    @user.build_bill_address(country_id: country_id) unless @user.bill_address
+
+    render :edit_address
+  end
+
+  def update_address
+    if @user.update_attributes(params[:user])
+      redirect_to spree.account_url, :notice => t(:account_updated)
```
Comment (mkllnk) [inline app/controllers/spree/users_controller_decorator.rb]: Old Ruby syntax here as well. And did you try the relative translation key `.account_updated` here? I'm not sure how it works with notices. I sometimes use the relative key and then look at the error message in the front end: "translation key ... missing". So I know where Rails tries to find the translation.

File: spec/features/consumer/account_spec.rb [hunk 6]
```diff
@@ -60,4 +60,34 @@
...
+    it 'updates shipping address' do
+      fill_in('user[ship_address_attributes][firstname]', with: 'John')
+      fill_in('user[ship_address_attributes][lastname]', with: 'Doe')
+      fill_in('user[ship_address_attributes][address1]', with: '11 Lovely Street')
+      fill_in('user[ship_address_attributes][phone]', with: '123-456-7890')
+      fill_in('user[ship_address_attributes][city]', with: 'Merbourne')
+      fill_in('user[ship_address_attributes][zipcode]', with: '3127')
+
+      click_button('Update')
+      expect(page).to have_content 'My Addresses'
```
Comment (mkllnk) [inline spec/features/consumer/account_spec.rb]: Yes. That should do the trick.

File: app/controllers/spree/users_controller_decorator.rb (hunk 4 above)
Comment (bingxie) [inline app/controllers/spree/users_controller_decorator.rb]: Good idea, if you run rubocop through the whole project there so many offenses. How did you check you code?

File: app/controllers/spree/users_controller_decorator.rb (hunk 5 above)
Comment (bingxie) [inline app/controllers/spree/users_controller_decorator.rb]: I just removed this notice, seems like on the front end page, there is no place to show it.

File: app/views/spree/users/edit_address.html.haml [hunk 7]
```diff
@@ -0,0 +1,15 @@
+.darkswarm
+  .row
+    = render partial: 'spree/shared/error_messages', locals: { target: @user }
+  .row
+    = form_for Spree::User.new, :as => @user, :url => main_app.account_update_address_path, :method => :put do |f|
```
Comment (mkllnk) [inline app/views/spree/users/edit_address.html.haml]: More old syntax here. :-)

File: app/views/spree/users/edit_address.html.haml [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+      .small-12.medium-6.columns
+        %h3.text-center= t('.edit_billing_address')
+        = render partial: 'spree/users/address_form', locals: { f: f, address_type: :bill_address }
+
+      .small-12.medium-6.columns
+        %h3.text-center= t('.edit_shipping_address')
+        = render partial: 'spree/users/address_form', locals: { f: f, address_type: :ship_address }
+
+      .medium-12.columns.text-center
+        = f.submit t(:update), :class => 'button primary'
```
Comment (mkllnk) [inline app/views/spree/users/edit_address.html.haml]: And here.

File: config/routes.rb [hunk 9]
```diff
@@ -48,6 +48,9 @@
     end
   end
 
+  get '/account/edit_address', :to => 'spree/users#edit_address'
```
Comment (mkllnk) [inline config/routes.rb]: And here.

Comment (bingxie) [inline config/routes.rb]: hi @mkllnk, after you think the code is good, then can you help me to deploy it on staging and let @sstead have a look on Wednesday?

**Comment by sstead** [comment]
Address isn't carrying through to the back end....
Default billing/shipping addresses are set on a shop by shop basis. When a customers saves an address in their account, it doesn't overwrite the address they've saved 'as default' during checkout, or that the enterprise has saved... It won't even populate the address details for a customer who's on a shop's customer list with no saved address. What's the point of this function if the address saved by the customer doesn't carry through to their shop's lists?

image
image

Idea...?  I think if a customer updates their address here, it should update as such in all shop customer lists that they are on, and all future shops they shop in. Then we should remove the 'save as default' checkbox in checkout.

I think a 'billing address same as shipping address' tick box would solve this...
image

**Comment by daniellemoorhead** [comment]
---Doing a PR Cleanup---

@oeoeaio @mkllnk what should we do with this PR from Bing? It was related to Standing Orders work, wondering if it's needed/replaced/still relevant?

**Comment by oeoeaio** [comment]
I think still relevant, but perhaps not 100% vital for the initial release. Will revisit once we've release the first cut of standing orders.

**Comment by myriamboure** [comment]
The connected issue #1132 is in the [Subscription project](https://github.com/openfoodfoundation/openfoodnetwork/projects/15) backlog, so I propose to close that PR, and when a dev pick the issue in the backlog to work on it, he will reopen it. @oeoeaio @mkllnk if you have any objection please reopen.

**Pull request closed without merging**